### PR TITLE
fix(dashboard): align validation error messages across the application

### PR DESCRIPTION
**Pull request by LetItRock**
### What changed? Why was the change needed?

Align the form validation error message across the application.

### Screenshots

Screenshot 2024-11-19 at 17 32 46

Screenshot 2024-11-19 at 17 33 22

Screenshot 2024-11-19 at 17 39 44

Screenshot 2024-11-19 at 17 41 21

### CODE REVIEW

File: apps/dashboard/src/components/create-workflow-button.tsx [hunk 1]
```diff
@@ -126,7 +134,7 @@ export const CreateWorkflowButton = (props: CreateWorkflowButtonProps) => {
                   <FormItem>
                     <FormLabel>Name</FormLabel>
                     <FormControl>
-                      <InputField>
+                      <InputField state={errors['name']?.message ? 'error' : 'default'}>
```
Comment (desiprisg) [inline apps/dashboard/src/components/create-workflow-button.tsx]: Are we going to do this for every field ?

File: apps/dashboard/src/components/header-navigation/edit-bridge-url-button.tsx [hunk 2]
```diff
@@ -86,7 +86,7 @@ export const EditBridgeUrlButton = () => {
                     <FormItem>
                       <FormLabel>Bridge Endpoint URL</FormLabel>
                       <FormControl>
-                        <InputField state={errors.bridgeUrl?.message ? 'error' : 'default'}>
+                        <InputField>
```
Comment (LetItRock) [inline apps/dashboard/src/components/header-navigation/edit-bridge-url-button.tsx:89]: Removed the `state` prop from the `InputField.` Under the hood, it determines that based on the form error state.

File: apps/dashboard/src/components/primitives/input.tsx [hunk 3]
```diff
@@ -59,16 +60,28 @@ const inputFieldVariants = cva(
...
   typeof inputFieldVariants
 >;
 
-const InputField = React.forwardRef<HTMLInputElement, InputFieldProps>(({ children, className, size, state }, ref) => {
-  return (
-    <div ref={ref} className={cn(inputFieldVariants({ size, state }), className)}>
-      {children}
-    </div>
-  );
+const InputFieldPure = React.forwardRef<HTMLInputElement, InputFieldPureProps>(
```
Comment (LetItRock) [inline apps/dashboard/src/components/primitives/input.tsx:67]: It is used for the fields that are not under the form.

File: apps/dashboard/src/components/primitives/tag-input.tsx [hunk 4]
```diff
@@ -55,7 +58,7 @@ const TagInput = forwardRef<HTMLInputElement, TagInputProps>((props, ref) => {
   return (
     <Popover open={isOpen}>
       <Command loop>
-        <div className="flex flex-col gap-2">
+        <div className="flex flex-col gap-2 pb-0.5">
```
Comment (LetItRock) [inline apps/dashboard/src/components/primitives/tag-input.tsx:61]: small padding for the error message to align it with the other fields that show errors

File: apps/dashboard/src/components/workflow-editor/schema.ts [hunk 5]
```diff
@@ -90,12 +89,12 @@ export const buildDynamicFormSchema = ({
     const isRequired = requiredFields.includes(key);
     let zodValue: z.ZodString | z.ZodNumber | z.ZodOptional<z.ZodString | z.ZodNumber>;
     if (value.type === 'string') {
-      zodValue = z.string().min(1, `${capitalize(key)} is required`);
+      zodValue = z.string().min(1);
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/schema.ts:92]: the default error messages are handled in the `overrideZodErrorMap` helper function

File: apps/dashboard/src/components/create-workflow-button.tsx (hunk 1 above)
Comment (LetItRock) [reply apps/dashboard/src/components/create-workflow-button.tsx]: updated the code to make our lives easier ;)

File: apps/dashboard/src/components/create-workflow-button.tsx [hunk 6]
```diff
@@ -170,6 +173,7 @@ export const CreateWorkflowButton = (props: CreateWorkflowButtonProps) => {
                     <FormControl>
                       <Textarea
                         placeholder="Description of what this workflow does"
+                        state={errors['description']?.message ? 'error' : 'default'}
```
Comment (rifont) [inline apps/dashboard/src/components/create-workflow-button.tsx]: Should we take a similar approach for `Textarea` primitive component state encapsulation as we're doing for the `InputField` [here](https://github.com/novuhq/novu/pull/7044/files#r1848710999)?

File: apps/dashboard/src/components/workflow-editor/schema.ts [hunk 7]
```diff
@@ -114,7 +113,7 @@ export const buildDynamicFormSchema = ({
       try {
         return JSON.parse(str);
       } catch (e) {
-        ctx.addIssue({ code: 'custom', message: 'Invalid payload. Payload needs to be a valid JSON.' });
+        ctx.addIssue({ code: 'custom', message: 'Payload needs to be a valid JSON' });
```
Comment (rifont) [inline apps/dashboard/src/components/workflow-editor/schema.ts]: Makes for easier reading with less words, also aligning "must" usage with other places.

```suggestion
        ctx.addIssue({ code: 'custom', message: 'Payload must be valid JSON' });
```

File: apps/dashboard/src/utils/validation.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import { z } from 'zod';
 import { capitalize } from './string';
 
-const getIssueField = (issue: z.ZodIssueBase) => capitalize(issue.path.join(' '));
+const getIssueField = (issue: z.ZodIssueBase) => capitalize(`${issue.path[issue.path.length - 1]}`);
```
Comment (rifont) [inline apps/dashboard/src/utils/validation.ts:4]: Nice catch 👍

File: apps/dashboard/src/utils/validation.ts [hunk 9]
```diff
@@ -21,7 +22,6 @@ const customErrorMap: z.ZodErrorMap = (issue, ctx) => {
     }
 
     if (issue.type === 'string') {
-      console.log({ issue });
```
Comment (rifont) [inline apps/dashboard/src/utils/validation.ts:24]: 🙏

File: apps/dashboard/src/components/create-workflow-button.tsx (hunk 6 above)
Comment (LetItRock) [reply apps/dashboard/src/components/create-workflow-button.tsx]: yes, we can do that and have `Textarea,` which would be used under the form, and have another `TextareaPure` for the cases when we don't have a form

File: apps/dashboard/src/utils/validation.ts [hunk 10]
```diff
@@ -70,6 +71,18 @@ const customErrorMap: z.ZodErrorMap = (issue, ctx) => {
...
+    return {
+      message: `${issueField} must be a valid email`,
+    };
+  }
+
+  if (issue.code === z.ZodIssueCode.invalid_enum_value) {
+    return {
+      message: `${issueField} must be one of ${issue.options.join(', ')}`,
+    };
+  }
```
Comment (rifont) [inline apps/dashboard/src/utils/validation.ts:85]: 💯

**Review by rifont — APPROVED** [review]
Awesome! 💯 ✨

File: apps/dashboard/src/components/create-workflow-button.tsx (hunk 6 above)
Comment (rifont) [reply apps/dashboard/src/components/create-workflow-button.tsx]: That sounds great!

File: apps/dashboard/src/components/primitives/input.tsx (hunk 3 above)
Comment (SokratisVidros) [reply apps/dashboard/src/components/primitives/input.tsx:67]: We need a better name for this. I am not happy with my suggestion. 

```suggestion
const InputFieldNaked = React.forwardRef<HTMLInputElement, InputFieldPureProps>(
```

File: apps/dashboard/src/components/primitives/tag-input.tsx [hunk 11]
```diff
@@ -20,7 +20,10 @@ const TagInput = forwardRef<HTMLInputElement, TagInputProps>((props, ref) => {
   const [tags, setTags] = useState<string[]>(value);
   const [inputValue, setInputValue] = useState('');
   const [isOpen, setIsOpen] = useState(false);
-  const validSuggestions = useMemo(() => suggestions.filter((suggestion) => !tags.includes(suggestion)), [tags]);
+  const validSuggestions = useMemo(
+    () => suggestions.filter((suggestion) => !tags.includes(suggestion)),
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/tag-input.tsx:24]: Maybe we can use a Set for this memo as an extra safe guard for tag uniquess.

**Review by SokratisVidros — APPROVED** [review]

**Review by BiswaViraj — APPROVED** [review]

**Pull request merged**